First steps for generating nuget packages for ARM64/Android.

cydhaselton opened the pull request:
Adds RIDS and related conditionals/variables to the appropriate files as part of Enabling Android/ARM64 for .NET Core ([issue here](https://github.com/dotnet/core-setup/issues/1651)).

Code appears to have been refactored since the original PR to which I was referred, cc-ing @janvorli and @qmfrederik for review in case of missed items

cydhaselton commented:
Also tagging @gkhanna79 for review because of #6056 and #6000

gkhanna79 commented on src/.nuget/Microsoft.NETCore.Native/Microsoft.NETCore.Native.props:
```diff
@@ -1,6 +1,6 @@
 <Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <SupportedPackageOSGroups>Linux;OSX</SupportedPackageOSGroups>
+    <SupportedPackageOSGroups>Linux;Android;OSX</SupportedPackageOSGroups>
```
Is there a reason you are enabling this here?

CC @jkotas

cydhaselton replied:
Because Linux and OSX were enabled here. If it's not applicable I can remove; I'm basically working from PRs and extrapolating based on refactored code so some of the edits may not be needed. Feel free to let me know and I'll make changes

gkhanna79 replied:
I think you can skip this file from the change.

gkhanna79 commented on src/.nuget/dir.props:
```diff
@@ -23,7 +23,7 @@
     <RuntimeOS Condition="'$(RuntimeOS)' == ''">$(OSRid)</RuntimeOS>
 
-    <SupportedPackageOSGroups Condition="'$(SupportedPackageOSGroups)' == ''">Windows_NT;OSX;Linux</SupportedPackageOSGroups>
+    <SupportedPackageOSGroups Condition="'$(SupportedPackageOSGroups)' == ''">Windows_NT;OSX;Android;Linux</SupportedPackageOSGroups>
```
What is the __DistroRid for Android build?

cydhaselton replied:
android.21-arm64 and/or android.21-arm, unless there's a difference between RID as registered in CoreFX and __DistroRid?

gkhanna79 replied:
Was this updated?

cydhaselton replied:
Not yet; still working out where __DistroRid originates and/or is generated (also see question below)

cydhaselton replied:
@gkhanna79 The test hack I added to build.sh generates __DistroRid for Android build and, with following commits, allows package build for Android.

I've added a TODO for figuring out a reliable way to generate __DistroRid for Android. Since I generated the commit changes from the `runtime.OS._filename_.props files and guesswork, a once-over to make sure I haven't missed anything would be appreciated

cydhaselton commented:
@jkotas and/or @gkhanna79:  Is there a definitive list of files to which a new RID should be added and/or detected for nuget package support?

cydhaselton commented:
@gkhanna79 The test hack I added to build.sh generates __DistroRid for Android build and, with following commits, allows package build for Android.

I've added a TODO for figuring out a reliable way to generate __DistroRid for Android. Since I generated the commit changes from the `runtime.OS._filename_.props files and guesswork, a once-over to make sure I haven't missed anything would be appreciated

cydhaselton commented:
Copying @mellinoe for additional visibility into package generation

cydhaselton commented:
cc-ing @qmfrederik @mellinoe and @janvorli for additional visibility into package generation

cydhaselton commented:
Also copying @schellap for input

gkhanna79 commented on build.sh:
```diff
@@ -74,7 +74,9 @@ initTargetDistroRid()
-                export __DistroRid=""
+                #Test hack to generate __DistroRid for Android
+                #TODO Find reliable way to generate __DistroRid for Android.
+                export __DistroRid="android.21-arm64"
```
Just to be clear - we cannot commit with this change. Have you considered specifying this at command (e.g. __DistroRid=... ./build.sh) and then adding a check in this function if __DistroRid is already defined or not?

cydhaselton replied:
@gkhanna79:  I did not, but that is a good idea.

jyoungyun replied:
IMHO, you need to save `ANDROID_PLATFORM` variable as a file when you build rootfs and then you can use it in building coreclr. :)

cydhaselton replied:
@jyoungyun This may be better; so far I've tried every iteration of detecting `__DistroRid` in this function and all have failed except for `! [[ -n "$__DistroRid" ]]`, which is the same as hardcoding that value.

cydhaselton replied:
@gkhanna79 See recent commit

cydhaselton replied:
FYI, Walking back all changes to build.sh and going route @jyoungyun  recommended. Too difficult getting logic check in `initTargetDistroRid()` to work without causing CI errors

gkhanna79 commented on src/.nuget/Microsoft.NETCore.Native/Microsoft.NETCore.Native.props:
```diff
@@ -3,4 +3,4 @@
   </PropertyGroup>
-</Project>
\ No newline at end of file
+</Project>
```
Undo change to this file.

cydhaselton replied:
@gkhanna79 Done

gkhanna79 commented on src/.nuget/Microsoft.NETCore.Runtime.CoreCLR/runtime.Android.Microsoft.NETCore.Runtime.CoreCLR.props:
```diff
@@ -0,0 +1,22 @@
+  </PropertyGroup>
+  <ItemGroup>
+    <NativeBinary Include="$(BinDir)libcoreclr.so" />
+    <NativeBinary Condition="'$(_PlatformDoesNotSupportNiFiles)' != 'true'" Include="$(BinDir)libcoreclrtraceptprovider.so" />
```
This check is incorrect. __PlatformDoesNotSupportNiFiles_ should be used for platforms where we cannot generate NI images right now. They just happen to be same as where this binary does not build.

If you are able to build this for Android, please fix this check accordingly so that it can be included for that platform.

FYI - @hqueue @hseok-oh @jyoungyun

gkhanna79 commented on src/.nuget/dir.props:
```diff
@@ -149,6 +162,14 @@
+    <OfficialBuildRID Include="android.21-arm64">
+      <Platform>arm64</Platform>
+    </OfficialBuildRID>
+    <OfficialBuildRID Include="android.21-arm">
```
Have you tested the build for Arm32 for Android? If so, your change for __DsitroRid should account for it as well.

cydhaselton replied:
I've only tested...and am only able to test...for arm64.  I can add arm into this change but won't be able to test it.

gkhanna79 replied:
In that case, please undo the Android for Arm change.

cydhaselton replied:
Removed

jyoungyun commented on config.json:
```diff
@@ -327,7 +327,7 @@
       "description": "Specifies the OS to build Core_Root for",
       "valueType": "property",
-      "values": [ "debian.8-x64", "fedora.23-x64", "opensuse.42.1-x64", "osx.10.12-x64", "rhel.7-x64", "ubuntu.14.04-x64", "ubuntu.16.04-x64", "ubuntu.16.10-x64" ],
+      "values": [ "debian.8-x64", "fedora.23-x64", "opensuse.42.1-x64", "osx.10.10-x64", "rhel.7-x64", "ubuntu.14.04-x64", "ubuntu.16.04-x64", "ubuntu.16.10-x64", "android.21-arm64", "android.21-arm" ],
```
I think you need to undo osx version change.

cydhaselton replied:
That's odd…don't remember making that change. Fixed

cydhaselton commented:
Tagging @qmfrederik to review/ok addition to `build-android-rootfs.sh`

cydhaselton commented:
General question: I'm still getting up to speed with CI builds. Is there anything I need to do with the previous commits in this PR that have failed build checks if a later commit or commits are successful?

qmfrederik commented:
@cydhaselton The change to `build-android-rootfs` is fine with me.

Regarding to the CI builds, only the last commit matters. If the previous commits are no longer relevant, you can squash them, and they will disappear from the list of commits in this pull request.

Would it be easier to treat Android as a Linux distribution? That would avoid the need to add a new OS group.

cydhaselton commented:
Another (probably obvious) question: is there a way for me manually to re-run a CI build for a specific commit?

qmfrederik commented:
@cydhaselton You can force the CI build to re-run, but only for the last commit in your PR.

cydhaselton commented:
@gkhanna79 @jyoungyun and others…any other changes/additions I should make to this PR before merge?

jyoungyun commented on build.sh:
```diff
@@ -73,13 +73,18 @@ initTargetDistroRid()
                 export __DistroRid="$ID.$VERSION_ID-$__BuildArch"
             fi
-        fi
+         fi
```
Remove redundant space :)

cydhaselton replied:
Good catch!  Fixed.

gkhanna79 commented on build.sh:
```diff
@@ -73,8 +73,13 @@ initTargetDistroRid()
-                export __DistroRid=""
+                if [ -e $ROOTFS_DIR/android_platform ]; then
+                    source $ROOTFS_DIR/android_platform
+                    export __DistroRid="$RID"
```
I presume that $RID is being set by the source statement earlier,right? What is the value it gets?

gkhanna79 replied:
Nevermind - saw it further below.

gkhanna79 commented on config.json:
```diff
@@ -327,7 +327,7 @@
       "description": "Specifies the OS to build Core_Root for",
       "valueType": "property",
-      "values": [ "debian.8-x64", "fedora.23-x64", "opensuse.42.1-x64", "osx.10.12-x64", "rhel.7-x64", "ubuntu.14.04-x64", "ubuntu.16.04-x64", "ubuntu.16.10-x64" ],
+      "values": [ "debian.8-x64", "fedora.23-x64", "opensuse.42.1-x64", "osx.10.12-x64", "rhel.7-x64", "ubuntu.14.04-x64", "ubuntu.16.04-x64", "ubuntu.16.10-x64", "android.21-arm64", "android.21-arm" ],
```
Were you not planning to remove the "arm" entry?

cydhaselton replied:
I did, but reverted when working on fixing initTargetDistroRid logic in build.sh and forgot to re-remove.
Fixed

gkhanna79 commented on src/.nuget/Microsoft.NETCore.ILAsm/runtime.Android.Microsoft.NETCore.ILAsm.props:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Thinking a bit more about this, why do you need to add these android specific files? Isn't the Linux version sufficient?

cydhaselton replied:
I'm not sure; I'm still piecing together how nuget packages are generated. Since I don't know all of the places a) where it would be important to separate Android from Linux and b) where it wouldn't matter, I'm taking a "separate everything now, integrate with Linux after" approach.

I can remove the files and see if package build still works.

cydhaselton replied:
@gkhanna79: So the build is successful, including packages, once `packageTargetOsGroup` for Android is removed from `dir.props` and conditionals added to `Microsoft.NETCore.Runtime.CoreCLR.pkgproj`

So my questions are:

1. What elements distinguish one package group from another? Or: What factors determine the creation of a new OS group?
2. Is it possible that a separate package group would be needed for Android in the future?

gkhanna79 replied:
Given that Android is sufficiently distinct, but still Linux based, I am fine if a new OS group for it is used while leveraging the existing Linux definitions of the components.

CC @chcosta

cydhaselton replied:
Unfortunately, unless I'm missing something, `packageTargetOSGroup` means adding Android-specific runtime-prefixed files (`runtime.Android...`).

cydhaselton replied:
cc @gkhanna79

chcosta replied:
I think it's reasonable for Android to use the Linux `_packageTargetOSGroup` to import the runtime package imports

cydhaselton replied:
@chcosta and @gkhanna79: Still not sure how to achieve what you are suggesting; if I create a `packageTargetOSGroup` for Android without adding the `runtime.Android` files package build fails.  Is there another OS that does this that I could reference?

gkhanna79 commented on src/.nuget/dir.props:
```diff
@@ -77,6 +78,13 @@
     </When>
+    <When Condition="'$(_runtimeOSFamily)' == 'android'">
+      <PropertyGroup>
+        <PackageRID>android.21-$(ArchGroup)</PackageRID>
```
Should we have a check for arm64 since that is all we are supporting for now?

cydhaselton commented:
@gkhanna79 @qmfrederik  and others; re-posting this question in general discussion for greater visibility. Code discussion [here](https://github.com/dotnet/coreclr/pull/10286#pullrequestreview-30691338):

So my questions are:
1. What elements distinguish one package group from another? Or: What factors determine the creation of a new OS group?
2. Is it possible that a separate package group would be needed for Android in the future?

cydhaselton commented:
cc @hseok-oh and @ericstj for insight into OS Group package info

danmosemsft commented:
@dotnet-bot test OSX10.12 x64 Checked Build and Test  (network)

cydhaselton commented:
Also cc-ing @jkotas as per [documentation](https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/contributing-workflow.md); is there a person or persons that are responsible for the packaging subsystem that I should copy on this for feedback on the OS Group question?

jkotas commented:
You have the right people copied on this already  (@gkhanna79, @ericstj).

cydhaselton commented:
Thanks, @jkotas. Would it be appropriate for me to update the CODE_OWNERS.TXT with that info?

jkotas commented:
Yes, it sounds good to me.

cydhaselton commented:
Polite ping to @ericstj and @gkhanna79 again for input into above questions

gkhanna79 commented on src/.nuget/Microsoft.NETCore.Runtime.CoreCLR/runtime.Android.Microsoft.NETCore.Runtime.CoreCLR.props:
```diff
@@ -0,0 +1,24 @@
+  </PropertyGroup>
+  <ItemGroup>
+    <NativeBinary Include="$(BinDir)libcoreclr.so" />
+    <NativeBinary Condition="'$(_PlatformDoesNotSupportNiFiles)' != 'true'" Include="$(BinDir)libcoreclrtraceptprovider.so" />
```
This just got recently changed to be supported. Thus, keeping it unified with the Linux definition will help streamline where it needs to be enabled vs not.

cydhaselton replied:
Should I revert this change then or keep it as is with the separate `PlatformDoesNotUseNiFiles` condition?

cydhaselton replied:
cc @gkhanna79

danmosemsft replied:
@gkhanna79 for q above.

chcosta commented on src/.nuget/dir.props:
```diff
@@ -34,6 +34,7 @@
   <!-- derive an OS Group based on the OS Family -->
   <PropertyGroup>
     <_derivedPackageTargetOSGroup Condition="'$(_derivedPackageTargetOSGroup)' == '' and '$(_runtimeOSFamily)' == 'osx'">OSX</_derivedPackageTargetOSGroup>
+    <_derivedPackageTargetOSGroup Condition="'$(_derivedPackageTargetOSGroup)' == '' and '$(_runtimeOSFamily)' == 'android'">Android</_derivedPackageTargetOSGroup>
```
@cydhaselton, `_packageTargetOSGroup` is the property which handles the runtime package import.  If you deleted this line, then I'd expect it to default to importing "Linux" runtime package projects.  That's probably not quite what we want though because it would then be ignoring the "Android" piece of the `SupportedPackageOSGroups`.  Instead, I would leave this code as-is, but add a conditional property definition here https://github.com/dotnet/coreclr/blob/8e986affba47efc1ce85ab0bb9428a6c28dae279/src/.nuget/dir.props#L47, which is something like

```
<_packageTargetOSGroup Condition="'$(_derivedPackageTargetOSGroup)' == 'Android'">Linux</_packageTargetOSGroup>
```

and maybe add a comment explaining that Android is using the same runtime pkgproj's as Linux.
```

cydhaselton replied:
@chcosta FYI, Implementing the above causes the package build for `Microsoft.NETCore.Runtime.CoreCLR` to fail. Specifically the build ignores the conditional I've copied from the `runtime.Android` file to the `runtime.Linux`file  to exclude `System.Private.Corelib.Ni.dll` from the package. If I use the existing condition for Ni  it does work…but since `Ni` files can be generated post-build on Android it shouldn't be used.

cydhaselton replied:
Polite ping to @gkhanna79 and @ericstj  for input into above question.

gkhanna79 replied:
@cydhaselton SPC.ni.dll generation has changed in the last week - can you please rebase and see if you are still running into this issue?

cydhaselton replied:
@gkhanna79 I am not; rebase resolved the issue. See updated code

cydhaselton commented:
@gkhanna79 The above looks messy; did you mean "merge" instead of "rebase?

gkhanna79 commented:
What is looking messy?

cydhaselton commented:
> What is looking messy?
@gkhanna79 Ignore…see edited message. Basically didn't do `git push -f` after rebase.

cydhaselton commented:
@dotnet-bot test Windows_NT x64 Debug Build and Test

cydhaselton commented:
@gkhanna79, @ericstj and others…is there anything else that needs to be done for this PR?

ericstj commented on config.json:
```diff
@@ -345,7 +345,7 @@
       "description": "Specifies the OS to build Core_Root for",
       "valueType": "property",
-      "values": [ "debian.8-x64", "fedora.24-x64", "fedora.25-x64", "opensuse.42.1-x64", "osx.10.12-x64", "rhel.7-x64", "ubuntu.14.04-x64", "ubuntu.16.04-x64", "ubuntu.16.10-x64" ],
+      "values": [ "debian.8-x64", "fedora.23-x64", "fedora.25-x64", "opensuse.42.1-x64", "osx.10.12-x64", "rhel.7-x64", "ubuntu.14.04-x64", "ubuntu.16.04-x64", "ubuntu.16.10-x64", "android.21-arm64" ],
```
This change to fedora appears to be a bad merge.

cydhaselton replied:
Crap. Fixed

ericstj commented on src/.nuget/dir.props:
```diff
@@ -102,6 +111,13 @@
+    <When Condition="'$(_runtimeOSFamily)' == 'android'">
+      <PropertyGroup>
+        <LibraryFileExtension>.so</LibraryFileExtension>
+<!--symbols included in .so, like Linux, but can be generated externally and if so, uses .debug ext-->
```
nit: indentation

cydhaselton replied:
Fixed

ericstj commented:
I've reviewed and commented on stuff that stuck out to me.  I don't see anything else concerning.  Most of these changes are around how coreclr decides to factor its package source which I don't have an opinion on so long as it produces the correct results.  I'd like for @gkhanna79 or @chcosta to give explicit sign off on that.

cydhaselton commented:
Thanks Eric.
@gkhanna79 and/or @chcosta let me know if there's anything you see that needs to be fixed.

cydhaselton commented:
Polite re-ping to @gkhanna79 and @chcosta … is there anything I need to change in this PR before merge?

chcosta commented on cross/build-android-rootfs.sh:
```diff
@@ -136,7 +137,11 @@ dpkg -x $__Android_Cross_Dir/deb/libunwind_1.2.20170304_$__AndroidArch.deb $__An
+# TODO: replace with logic that generates RID based on $__ApiLevel and $__BuildArch
+echo "Generating platform file..."
 
+echo "RID=android.21-arm64" > $__ToolchainDir/sysroot/android_platform
```
Any reason not to address this TODO and use the ApiLevel / BuildArch here?

cydhaselton replied:
@chcosta The TODO was addressed in previous commit by adding logic to `build-android-rootfs.sh`; I accidentally left the comment. Will remove comment and echo commands.

chcosta commented on src/.nuget/dir.props:
```diff
@@ -163,6 +179,11 @@
     </OfficialBuildRID>
   </ItemGroup>
+  <ItemGroup Condition="$(SupportedPackageOSGroups.Contains(';Android;'))">
+    <OfficialBuildRID Include="android.21-arm64">
```
Do you want to include the Portable rid?

cydhaselton replied:
@chcosta At this line? I don't think so…a quick check of master shows that the other supported OSs don;t include it at this point. Should it be?

@gkhanna79 Anything you see that needs changing?

gkhanna79 replied:
OfficialBuildRID is only used for platform targets that the official .NET Core build is going to support and we should add it only when we enable building for it. Otherwise, this is not required.

@cydhaselton Can you please undo this change for now?

cydhaselton replied:
Out of town but will push change when I return.

cydhaselton replied:
@gkhanna79 Removed

chcosta commented:
Two additional comments, and then LGTM

cydhaselton commented:
@dotnet-bot test Tizen armel Cross Debug Build and Windows_NT x64 Debug Build and Test

cydhaselton commented:
@dotnet-bot test Windows_NT x64 Debug Build and Test

cydhaselton commented:
@gkhanna79 and @chcosta are there any other changes needed for this PR?

cydhaselton commented:
FYI will be keeping this branch around until I have time to write up documentation.
